framework/src/systems/NonlinearSystem.C

```diff
@@ -410,3 +396,126 @@ NonlinearSystem::converged()
+      {
+        auto & field_variable = *field_variables[i];
+        dof_map.dof_indices(elem, dof_indices, field_variable.number());
+        for (auto dof_index : dof_indices)
+          if (dof_map.local_index(dof_index))
+          {
+            // For now we will use the diagonal for determining scaling
+            auto mat_value = diagonal(dof_index);
+            inverse_scaling_factors[i] = std::max(inverse_scaling_factors[i], std::abs(mat_value));
+          }
+      }
+    }
+
+    auto offset = field_variables.size();
+
+    // Compute scalar factors for scalar variables
+    for (MooseIndex(scalar_variables) i = 0; i < scalar_variables.size(); ++i)
+    {
+      auto & scalar_variable = *scalar_variables[i];
+      dof_map.SCALAR_dof_indices(dof_indices, scalar_variable.number());
+      for (auto dof_index : dof_indices)
+        if (dof_map.local_index(dof_index))
+        {
+          // For now we will use the diagonal for determining scaling
+          auto mat_value = diagonal(dof_index);
+          inverse_scaling_factors[offset + i] =
+              std::max(inverse_scaling_factors[offset + i], std::abs(mat_value));
+        }
+    }
+
+    // Get the maximum value across processes
+    _communicator.max(inverse_scaling_factors);
+
+    // We have to make sure that our scaling values are not zero
+    for (auto & scaling_factor : inverse_scaling_factors)
+      if (scaling_factor < std::numeric_limits<Real>::epsilon())
+        scaling_factor = 1;
+
+    if (_verbose)
+    {
+      _console << "Automatic scaling factors:\n";
+      auto original_flags = _console.flags();
+      auto original_precision = _console.precision();
+      _console.unsetf(std::ios_base::floatfield);
+      _console.precision(6);
+
+      for (MooseIndex(field_variables) i = 0; i < field_variables.size(); ++i)
+      {
+        auto & field_variable = *field_variables[i];
+        _console << "  " << field_variable.name() << ": " << 1.0 / inverse_scaling_factors[i]
+                 << "\n";
+      }
+      for (MooseIndex(scalar_variables) i = 0; i < scalar_variables.size(); ++i)
+      {
+        auto & scalar_variable = *scalar_variables[i];
+        _console << "  " << scalar_variable.name() << ": "
+                 << 1.0 / inverse_scaling_factors[offset + i] << "\n";
+      }
+      _console << "\n\n";
+
+      // restore state
+      _console.flags(original_flags);
+      _console.precision(original_precision);
+    }
+
+    // Now set the scaling factors for the variables
+    applyScalingFactors(inverse_scaling_factors);
+    if (auto displaced_problem = _fe_problem.getDisplacedProblem().get())
+      displaced_problem->systemBaseNonlinear().applyScalingFactors(inverse_scaling_factors);
+
+    // Now it's essential that we reset the sparsity pattern of the matrix
+    petsc_matrix.reset_preallocation();
+
+#else
+    mooseWarning("Automatic scaling requires a PETSc version of 3.9.0 or greater, so no automatic "
```

Comment:
why not just error out?

Reply:
If that is what an application developer would prefer